### ISSUE TITLE
AsyncClient stuck on many commands

### ISSUE THREAD

**Issue by thhave** [aerospike/aerospike-client-csharp#71]
Hello. 
We use AsyncClient to perform write operations on many keys using `TaskCompletionSource` (or `Channel<T>` from `System.Threading.Channels`) to wait for all keys to complete.

After updating the client to version 5.2.1 we encountered a strange behavior.

Consider the following test :
```csharp
    [Parallelizable(ParallelScope.None)]
    public class AsyncClientTest
    {
        private readonly Host[] _aerospikeHosts;

        public AsyncClientTest()
        {
            _aerospikeHosts = new Host[]
            {
                /* list of our aerospike hosts */
            };
        }

        [TestCase(MaxCommandAction.DELAY, 3000)]
        [TestCase(MaxCommandAction.DELAY, 2000)]
        public async Task PutLargeBatch_Test(MaxCommandAction maxCommandAction, int itemCount)
        {
            using var asyncClient = new AsyncClient(new AsyncClientPolicy
            {
                // some our defaults:
                asyncMaxCommandAction = maxCommandAction,
                useServicesAlternate = true,
                writePolicyDefault = new WritePolicy
                {
                    sendKey = true,
                    totalTimeout = 30000,
                    socketTimeout = 5000,
                    recordExistsAction = RecordExistsAction.CREATE_ONLY,
                    expiration = 600,
                    maxRetries = 0
                },
                asyncMaxCommands = 200
            }, _aerospikeHosts);

            using var cts = new CancellationTokenSource(60000);

            var data = Enumerable.Range(0, itemCount).Select(_ => new TestModel()).ToArray();

            var completionSource = new AerospikeOperationCompletionSource(data.Length, cts.Token);

            foreach (var item in data)
            {
                var key = new Key(ns: "stage", setName: "as_client_test", item.Id.ToString());

                var writeListener = new AerospikeWriteListener(key, completionSource);

                asyncClient.Put(null, writeListener, key, new Bin("id", item.Id.ToString()));
            }

            // wait for all keys to complete
            Func<Task> asyncAction = () => completionSource.Task;

            await asyncAction.Should().NotThrowAsync().ConfigureAwait(false);

            var results = await completionSource.Task.ConfigureAwait(false);

            results.Should().HaveCount(data.Length);

            results.Should().AllSatisfy(r => r.Success.Should().BeTrue());
        }

        public class AerospikeResult
        {
            private readonly AerospikeException _exception;

            public AerospikeResult(Key key, AerospikeException exception = null)
            {
                Argument.IsNotNull(key, nameof(key));

                Key = key;

                _exception = exception;
            }

            public Key Key { get; }

            public bool Success => _exception == null;
        }

        public class AerospikeOperationCompletionSource
        {
            private int _operationCount;
            private readonly TaskCompletionSource<AerospikeResult[]> _tcs;
            private readonly CancellationTokenRegistration _ctr;

            private readonly ConcurrentBag<AerospikeResult> _results;

            public AerospikeOperationCompletionSource(int operationCount, CancellationToken cancellationToken = default)
            {
                if (operationCount < 1)
                {
                    throw new ArgumentOutOfRangeException(nameof(operationCount));
                }

                _operationCount = operationCount;

                _tcs = new TaskCompletionSource<AerospikeResult[]>(TaskCreationOptions.RunContinuationsAsynchronously);
                _ctr = cancellationToken.Register(() => _tcs.TrySetCanceled(), useSynchronizationContext: false);

                _results = new ConcurrentBag<AerospikeResult>();
            }

            public Task<AerospikeResult[]> Task => _tcs.Task;

            public void OnOperationPerformed(AerospikeResult result)
            {
                Argument.IsNotNull(result, nameof(result));

                _results.Add(result);

                if (Interlocked.Decrement(ref _operationCount) == 0)
                {
                    _tcs.SetResult(_results.ToArray());
                }
            }

            public void Dispose()
            {
                _ctr.Dispose();
            }
        }

        public sealed class AerospikeWriteListener : WriteListener
        {
            private readonly Key _key;
            private readonly AerospikeOperationCompletionSource _completionSource;

            public AerospikeWriteListener(Key key, AerospikeOperationCompletionSource completionSource)
            {
                _key = key;
                _completionSource = completionSource;
            }

            public void OnSuccess(Key key)
            {
                _completionSource.OnOperationPerformed(new AerospikeResult(key));
            }

            public void OnFailure(AerospikeException exception)
            {
                _completionSource.OnOperationPerformed(new AerospikeResult(_key, exception));
            }
        }

        public class TestModel
        {
            public TestModel()
            {
                Id = Guid.NewGuid();
            }

            public Guid Id { get; set; }
        }
    }
```
The test fails if item count is greater than ~2000 items at client version 5.2.1 meaning that the client starts ignoring some queued keys and doesn't return any result on it (no exceptions).

But at 5.2.0 and earlier the test succeeds for 100000+ items with the same client policy configuration and normally can throw timeout exception if key count becomes too large to complete them quickly.

I tried to log operation completions to file (without using cancellation) and then I saw that after a few thousand completions nothing is happening (no invocations of listener's methods). Increasing total timeout doesn't change the behaviour.

Server version: aerospike:ce-6.0.0.1

**Comment by BrianNichols** [comment]
I have duplicated the error with C# client 5.2.1 and will investigate further.  I recommend staying on 5.2.0 until the investigation is complete.

I did notice that asyncMaxCommands was set to 200, but asyncMaxConnsPerNode was left at the new default (100).  I suggest setting  asyncMaxConnsPerNode to 200 as well.

**Comment by BrianNichols** [comment]
The error should be fixed in the new C# client 5.2.2:

https://download.aerospike.com/download/client/csharp/notes.html#5.2.2
